### PR TITLE
storage: fix reclocking correctness bug

### PR DESCRIPTION
**Pull request by petrosagg**
The reclock operator previously exposed an API that allowed the caller
to map the max observed offset to a reclocked timestamp. This API
allowed for non-definite behaviour in the face of restarts.

The way this would happen is that on the first run a `storaged` would
read data until some offset `X`, and produce some bindings, and then
upon restart another `storaged` instance would read data until some
offset `Y > X` and also try to produce bindings. Since `Y > X` a later
timestamp would be chosen but instead of reclocking only the messages
at offsets `X < offset <= Y` at this new timestamp all the data was
being pushed to the future.

The result is that duplicated data end up in the output.

This PR fixes the above by changing the relock operator API to reclock
individual messages intead of just the upper frontier and also breaks
down the various steps it performs in order to aid understanding and
allow unit testing the code.

### CODE REVIEW

File: src/storage/src/source/reclock.rs [hunk 1]
```diff
@@ -8,413 +8,668 @@
...
-                // We take the minimum of all of these as well as `read_progress` as we know no new bindings can have a
-                // time before that (because we read from the listener in ascending timestamp order).
-                let progress = self
-                    .persisted_timestamp_bindings
-                    .iter()
-                    .filter_map(|(partition, bindings)| {
-                        if let Some((ts, _offset)) = matched_offsets.get(partition) {
-                            return Some(*ts);
+        // Tail the listen stream until we reach the target upper frontier
+        while PartialOrder::less_than(&self.upper, target_upper) {
```
Comment (guswynn) [inline src/storage/src/source/reclock.rs:219]:
```suggestion
        // Tail the listen stream until we reach the target upper frontier. Note that, in the common case, we are also the writer, so we are waiting to read-back what we wrote
        while PartialOrder::less_than(&self.upper, target_upper) {
```

File: src/storage/src/source/reclock.rs [hunk 2]
```diff
@@ -8,413 +8,668 @@
...
+    }
+
+    /// Ensures that the persist shard backing this reclock operator contains bindings that cover
+    /// the provided source frontier by minting bindings where appropriate.
+    ///
+    /// When this function returns the local dTVC view of the remap collection will contain
+    /// definite timestamp bindings that can be used to timestamp data at these offsets.
+    async fn mint<P: Borrow<PartitionId>>(
+        &mut self,
+        source_frontier: &HashMap<P, MzOffset>,
```
Comment (guswynn) [inline src/storage/src/source/reclock.rs:248]: lets call it `source_upper` for consistency with `advance`

**Review by guswynn — COMMENTED** [review]
@cjubb39 probably wants to look at this, but looks good!

i have 1 question about compact, then a bunch of nits that are just adding clarifying comments

File: src/storage/src/source/reclock.rs [hunk 3]
```diff
@@ -8,413 +8,668 @@
...
-            self.write_upper = match compare_and_append_result {
-                Ok(()) => new_upper,
-                Err(Upper(actual_upper)) => actual_upper,
-            };
+        }
 
-            while PartialOrder::less_than(&self.read_progress, &self.write_upper) {
+        // If there no updates we might still want to bump the upper time if enough time has passed
+        let mint_timestamp = if updates.is_empty() {
+            self.next_mint_timestamp().now_or_never()
```
Comment (guswynn) [inline src/storage/src/source/reclock.rs:264]: `now_or_never` has confusing semantics, I would just make a separate ts function that doesnt do the sleeping

File: src/storage/src/source/reclock.rs [hunk 4]
```diff
@@ -8,413 +8,668 @@
...
-                            self.persisted_timestamp_bindings
-                                .entry(partition)
-                                .or_insert_with(VecDeque::new)
-                                .push_back((timestamp, diff));
-                        }
-                    }
+                    Ok(Ok(Ok(()))) => break,
+                    Ok(Ok(Err(actual_upper))) => return Err(actual_upper),
+                    Ok(Err(invalid_use)) => panic!("{invalid_use:?}"),
+                    Err(_external_err) => continue,
```
Comment (guswynn) [inline src/storage/src/source/reclock.rs:283]: `tracing::debug` the error at least

also, can we add a comment on why retrying is okay?

File: src/storage/src/source/reclock.rs [hunk 5]
```diff
@@ -8,413 +8,668 @@
...
+
+impl<'a, M> Iterator for ReclockPartIter<'a, M> {
+    type Item = (M, Timestamp);
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let (message, offset) = self.messages.next()?;
+        // Skip bindings whose source offset upper doesn't cover this message's offset
+        while !(offset < self.bindings.peek().expect("not enough bindings").1) {
+            self.bindings.next();
+        }
```
Comment (guswynn) [inline src/storage/src/source/reclock.rs:368]: note that this depends on `self.messages` being sorted

File: src/storage/src/source/reclock.rs [hunk 6]
```diff
@@ -8,413 +8,668 @@
...
+        );
+        let reclocked_msgs = operator
+            .reclock(&mut batch)
+            .await
+            .flat_map(|(_, msgs)| msgs)
+            .collect_vec();
+        assert_eq!(reclocked_msgs, &[(3, 2000), (4, 2000)]);
+        assert!(batch[&PART_ID].is_empty());
+
+        // Attempt to reclock offset 2 which should cause a panic
```
Comment (guswynn) [inline src/storage/src/source/reclock.rs:610]: if you improve the panic message in my suggestion above, we can make this test more reliable by making the expected panic message the full message with the erroneous offset

File: src/storage/src/source/reclock.rs [hunk 7]
```diff
@@ -8,413 +8,668 @@
...
+        const PART_ID: PartitionId = PartitionId::None;
+
+        // Create two operators pointing to the same shard
+        let shared_shard = ShardId::new();
+        let location = PersistLocation {
+            blob_uri: "mem://".to_owned(),
+            consensus_uri: "mem://".to_owned(),
+        };
+        let client = location.open().await.unwrap();
+        let mut op_a = test_operator_with_client(client.clone(), shared_shard).await;
```
Comment (guswynn) [inline src/storage/src/source/reclock.rs:630]: oh nice this works to share a memory location!

File: src/storage/src/source/reclock.rs [hunk 8]
```diff
@@ -8,413 +8,668 @@
...
+            vec![(1, MzOffset::from(1)), (2, MzOffset::from(2))],
+        );
+        let reclocked_msgs = op_a
+            .reclock(&mut batch)
+            .await
+            .flat_map(|(_, msgs)| msgs)
+            .collect_vec();
+        assert_eq!(reclocked_msgs, &[(1, 1000), (2, 1000)]);
+        assert!(batch[&PART_ID].is_empty());
+
```
Comment (guswynn) [inline src/storage/src/source/reclock.rs:648]: please also `advance` `op_a` here to show that the `op_b` code below still works

File: src/storage/src/source/reclock.rs [hunk 9]
```diff
@@ -8,413 +8,668 @@
...
-        // Mod and then round up
-        let remainder = new_ts % update_interval;
-        let new_ts_rounded = if remainder == 0 {
-            new_ts
-        } else {
-            new_ts - remainder + update_interval
-        };
+    /// Calling this method promises that all future timestamping requests will be for offsets
+    /// greater or equal to the provided source upper frontier. In exchange the reclock operator
+    /// return the timestamp frontier that the caller can use to downgrade its capability to.
```
Comment (cjubb39) [inline src/storage/src/source/reclock.rs:160]: super nit: `return` -> `returns`

File: src/storage/src/source/reclock.rs [hunk 10]
```diff
@@ -8,413 +8,668 @@
...
+        }
 
-        if new_ts_rounded > self.last_timestamp {
-            self.last_timestamp = new_ts_rounded;
-            Some(new_ts_rounded)
-        } else {
-            None
+        // Then, figure out the highest timestamp the caller we downgrade to, which is the minimum
+        // timestamp of all the bindings involved in the given frontier.
+        let mut handle_since = self.upper.clone();
```
Comment (cjubb39) [inline src/storage/src/source/reclock.rs:172]: I think there might be something implicit in the "what the caller can downgrade capability to" because it's not clear to me that `handle_since` is a valid return value of the offset in `source_upper` is beyond `self.last_source_upper` for all partitions.  Can you add a little more detail to the doc-comment on this function relating the return value to the timestamps that can be returned by `Self::relock` in the future?

File: src/storage/src/source/reclock.rs [hunk 11]
```diff
@@ -8,413 +8,668 @@
...
+        for pid in self.last_source_upper.keys() {
+            let offset = source_upper.get(pid).cloned().unwrap_or_default();
+            if let Some((ts, _)) = self
+                .partition_uppers(pid)
+                .find(|(_, upper)| offset < *upper)
```
Comment (cjubb39) [inline src/storage/src/source/reclock.rs:177]: Should this be `offset <= *upper`? I think we'd want to constrain the returned antichain if we're going to allow for more messages at the offset given by `source_upper[pid]` as they'd be reclocked at `ts`.

I don't _think_ there's a test that stresses multiple messages for a given offset but if I'm mistaken can you point me there? If they are in fact missing, can you add some?

File: src/storage/src/source/reclock.rs [hunk 12]
```diff
@@ -8,413 +8,668 @@
...
             }
+            self.sync(&new_upper).await;
+        }
+        Ok(())
+    }
 
-            debug_assert!(self.validate_persisted_bindings());
+    /// Produces a new timestamp suitable for minting bindings. The returned value is guaranteed to
+    /// be beyond the current `upper` frontier and a multiple of `update_interval_ms`.
+    async fn next_mint_timestamp(&mut self) -> Timestamp {
```
Comment (cjubb39) [inline src/storage/src/source/reclock.rs:293]: what do you think of adding a TODO about large amount of clock skew?

**Review by cjubb39 — COMMENTED** [review]
sorry for all the comments on documentation lol but I do have a couple correctness-related questions and clarifications inline, too!

File: src/storage/src/source/reclock.rs (hunk 12 above)
Comment (petrosagg) [reply src/storage/src/source/reclock.rs:293]: I tried to figure out how this operator would even know that skew had happened and couldn't. Maybe worth talking about it on slack if you have some ideas

File: src/storage/src/source/reclock.rs (hunk 5 above)
Comment (petrosagg) [reply src/storage/src/source/reclock.rs:368]: Indeed, and they are sorted in the `reclock` method

File: src/storage/src/source/reclock.rs (hunk 6 above)
Comment (petrosagg) [reply src/storage/src/source/reclock.rs:610]: Changed these to Results for better testing

File: src/storage/src/source/reclock.rs (hunk 8 above)
Comment (petrosagg) [reply src/storage/src/source/reclock.rs:648]: Done!

**Review by cjubb39 — COMMENTED** [review]
Two comments inline worth highlighting 1) about `as_of` in `partition_since`; 2) the `while let Err(_) = self.mint` loop.  The rest are pretty small.

Thanks for this round of changes -- I think it's easer to reason about and review!

File: src/storage/src/source/reclock.rs [hunk 13]
```diff
@@ -8,413 +8,668 @@
...
 
-        // This is an invariant of the persist collection.
         assert!(
-            timely::PartialOrder::less_equal(&read_progress, &write_upper),
-            "{:?} PARTIAL ORDER: {:?} {:?}",
-            name,
-            read_progress,
-            write_upper,
+            PartialOrder::less_equal(since, &as_of),
+            "invalid as_of: as_of({as_of:?}) < since({since:?})"
```
[the remark this answers is not in the document]
Comment (guswynn) [reply src/storage/src/source/reclock.rs:97]: Lets add a comment here? or even better, assert that the `as_of` is a single-element antichain?

File: src/storage/src/source/reclock.rs [hunk 14]
```diff
@@ -8,413 +8,668 @@
...
-                    "TS GOES BACKWARDS: {:?} -> {:?} in PART {:?}; BINDINGS: {:?}",
-                    last_ts,
-                    ts,
-                    partition,
-                    self.persisted_timestamp_bindings
-                );
-                last_ts = *ts;
+    /// Compacts the internal state
+    #[allow(dead_code)]
+    pub async fn compact(&mut self, new_since: Antichain<Timestamp>) {
```
[the remark this answers is not in the document]
Comment (guswynn) [reply src/storage/src/source/reclock.rs:212]: this is true, I forgot that the storage controller can request a source with an `as_of`, later on

File: src/storage/src/source/reclock.rs [hunk 15]
```diff
@@ -8,413 +8,795 @@
...
+            .reclock(&mut batch)
+            .await
+            .unwrap()
+            .flat_map(|(_, msgs)| msgs)
+            .collect_vec();
+        assert_eq!(reclocked_msgs, &[(1, 1000), (2, 1000)]);
+        assert!(batch[&PART_ID].is_empty());
+
+        // Also compact operator A. Since operator B has its own read handle it shouldn't affect it
+        op_a.compact(Antichain::from_elem(1000)).await;
```
Comment (guswynn) [inline src/storage/src/source/reclock.rs:773]: can you additionally call `reclock_frontier` on `op_a` here? It doesnt mutate the operator, but we can ensure that A is making progress as expected

**Review by guswynn — COMMENTED** [review]
I think all thats left is @cjubb39's very thoughtful comments!

File: src/storage/src/source/reclock.rs [hunk 16]
```diff
@@ -8,413 +8,795 @@
...
-                        VecDeque::from_iter(std::iter::once((ts, offset))),
-                    )
-                })
-                .collect()
-        } else {
-            HashMap::new()
-        };
-        let read_cursors = HashMap::new();
+        // Ensure we have enough bindings and re-sync state if we lost a race
+        while let Err(Upper(global_upper)) = self.mint(&batch_upper).await {
```
[the remark this answers is not in the document]
Comment (petrosagg) [reply src/storage/src/source/reclock.rs:155]: It sounds like in your example `R1` must always be making progress for `R0` to seemingly get stuck, but the system as a whole is making progress because the offsets did in fact get reclocked and not only that the frontier is constantly being pushed forward by `R1` (otherwise `R0` wouldn't be losing the races).

This means that an external observer of STORAGE (e.g COMPUTE) will be seeing the ingestion progressing normally. Does this match your understanding?

I think the real question is what should a `storaged` process do when it discovers a race. When a `compare_and_append` fails at the end of the pipeline where we write to the final persist collection we currently panic and crash. This enough to ensure that we don't end up with a bunch of `storaged` instances all competing with each other so it's unclear if we should do the same in all `compare_and_append` operation. One reason to *not* do it is that it is possible to race with yourself. If you to a `compare_and_append` that succeeds on the persist side but due to networks errors comes back as an `ExternalError` then we'll retry and fine that the upper has moved and initiate a sync. In the "raced with myself" scenario this is the correct thing to do and not to crash. Crashing however is also fine, the orchestrator will notice that we're out of storaged processes and schedule a new one shortly. Shrug.

File: src/storage/src/source/reclock.rs [hunk 17]
```diff
@@ -8,413 +8,795 @@
...
+            offset: MzOffset::default(),
+            bindings,
+        }
+    }
 
-            debug_assert!(self.validate_persisted_bindings());
+    /// Returns the since frontier for a given partition
+    fn partition_since(&self, pid: &PartitionId) -> MzOffset {
+        if self.since.elements() == [Timestamp::minimum()] {
+            // If we never compacted in the DestTime domain then the SourceTime domain isn't
```
[the remark this answers is not in the document]
Comment (petrosagg) [reply src/storage/src/source/reclock.rs:364]: The operator knows if a compaction has happened or not by checking if the since frontier is `Antichain::from_elem(Timestamp::minimum())` or not. When the operator is started with an `as_of` frontier then the since frontier is initialized as such because during the initial snapshot [persist will *compact itself*. ](https://github.com/MaterializeInc/materialize/blob/main/src/persist-client/src/read.rs#L138)

In other words there is no material difference between:
1. starting the operator `as_of 0` and then compacting to timestamp `t`
2. starting the operator `as_of t`

The final state is indistinguishable and both should be treated as compaction having happened.

I will add a test that uses `as_of` initialization nevertheless to exercise that code path too.

File: src/storage/src/source/reclock.rs [hunk 18]
```diff
@@ -8,413 +8,795 @@
...
+        // The upper is the greatest frontier that we can ever return
+        let mut dest_frontier = self.upper.clone();
+
+        let mut partitions = HashSet::new();
+        partitions.extend(self.source_upper.keys());
+        partitions.extend(source_frontier.keys());
+        // To refine it we have to go through all the partitions we know about and:
+        for pid in partitions {
+            let offset = source_frontier.get(pid).copied().unwrap_or_default();
+            // Ensure that the offsets are beyond the source since frontier
```
[the remark this answers is not in the document]
Comment (petrosagg) [reply src/storage/src/source/reclock.rs:187]: This is a tricky bit and your comment here means I didn't do a good job explaining it in the comments.

The piece that's missing from the explanation is that there is no such thing as "partition you didn't ask for". All partitions are to be considered always present in a source frontier. What I mean by that is that the frontiers `HashMap::new()` and `HashMap::from_iter([(PartitionId::Kafka(1), MzOffset::from(0)), (PartitionId::Kafka(2), MzOffset::from(0))])` are exactly the same thing. Ideally we wouldn't be using a `HashMap` to represent the source frontier and use an actual `Antichain`. But because the set of partitions is very large the `Antichain` would have to have billions of zero entries.

In other words, the source frontier `HashMap`s are to be thought of as a special `Antichain` that absence of `PartitionId` implies the minimum offset is the upper for that partition id.

Back to your original question:

> As I understand it, if there's a partition the operator knows about but you didn't ask for (including in source_frontier), this will return an error only if you've compacted.

If a partition is in the `HashMap` of the operator then it means that its since is some non-zero value, otherwise it would be absent. So if someone calls `reclock_frontier` with a `HashMap` that doesn't include that partition **they are actually asking to reclock offset zero**, which we cannot do because we've compacted past it. So it's correct to return an error in this case.


> We could also add something similar if there's a partition the operator doesn't know about but you ask for? 

This is also handled the same way. In this case the partition since the operator knows about will be assumed to be zero (since it's absent) and you're trying to reclock some later offset for which a binding doesn't exist. In this case the function correctly ignores this in the calculation because the original estimate `self.upper` is valid since when the time comes to mint that binding its timestamp will surely be beyond `self.upper`.

File: src/storage/src/source/reclock.rs [hunk 19]
```diff
@@ -8,413 +8,795 @@
...
+            // At this point we have successfully produced data in the reclock shard so we need to
+            // sync to update our local view as well
+            self.sync(&new_upper).await;
+        }
+        Ok(())
+    }
+
+    /// Produces a new timestamp suitable for minting bindings or the amount of time that the
+    /// caller needs to wait for one to become available. On success, the returned timestamp is
+    /// guaranteed to be beyond the current `upper` frontier and a multiple of `update_interval_ms`
```
[the remark this answers is not in the document]
Comment (petrosagg) [reply src/storage/src/source/reclock.rs:334]: @aljoscha that's right! We just need that `upper <= new_ts` holds

File: src/storage/src/source/reclock.rs [hunk 20]
```diff
@@ -8,413 +8,795 @@
...
 use mz_persist_client::write::WriteHandle;
-use mz_persist_client::Upper;
+use mz_persist_client::{PersistClient, Upper};
 use mz_repr::Timestamp;
-use tracing::{error, info};
 
+/// The reclock operator reclocks a stream that is timestamped with some timestamp `SourceTime`
+/// into another time domain that is timestamped with some timestamp `DestTime`.
+///
+/// Currently the `SourceTime` is hardcoded to `HashMap<PartitionId, MzOffset>`
```
[the remark this answers is not in the document]
Comment (petrosagg) [reply src/storage/src/source/reclock.rs:36]: oops, yes, that's right. I'll change the doc

File: src/storage/src/source/reclock.rs (hunk 4 above)
Comment (petrosagg) [reply src/storage/src/source/reclock.rs:322]: I added a comment explaining why retrying is ok. Correctness isn't relying on any particular pattern of calling `mint` or `sync`. You can call them however you want (mint without sync, sync to wrong frontiers then mint, etc) and you'll never get back bad bindings. You'll either get correctly minted bindings or an error or a hang (e.g if you sync past the upper).

Calling the methods in the way they are in this PR just ensures that we make forward progress

File: src/storage/src/source/reclock.rs [hunk 21]
```diff
@@ -8,413 +8,795 @@
...
+    fn partition_since(&self, pid: &PartitionId) -> MzOffset {
+        if self.since.elements() == [Timestamp::minimum()] {
+            // If we never compacted in the DestTime domain then the SourceTime domain isn't
+            // compated either. Therefore the since frontier is zero
+            MzOffset::default()
+        } else {
+            // If we have compacted there are two posibilities. Either the since frontier is at
+            // exactly the time of the first binding (and will continue to do so since times are
+            // advanced during compaction), or it is behind it.
+            let first_binding = self
```
[the remark this answers is not in the document]
Comment (petrosagg) [reply src/storage/src/source/reclock.rs:371]: good point, changed!

File: src/storage/src/source/mod.rs [hunk 22]
```diff
@@ -787,87 +786,71 @@ where
...
-                        if source_stream.is_done() {
-                            // We just emitted the last piece of data that needed to be timestamped
-                            if progress_some {
+                        for err in non_definite_errors.drain(..) {
+                            // TODO: make errors definite
+                            yield Event::Message(0, Err(err));
+                        }
+                        // TODO(petrosag): compaction should be driven by AllowCompaction commands
+                        // coming from the storage controller
+                        let new_ts_upper = timestamper.reclock_frontier(&source_upper).unwrap();
```
[the remark this answers is not in the document]
Comment (petrosagg) [reply src/storage/src/source/mod.rs:847]: fixed

File: src/storage/src/source/reclock.rs [hunk 23]
```diff
@@ -8,413 +8,795 @@
...
+        &'a mut self,
+        batch: &'a mut HashMap<PartitionId, Vec<(M, MzOffset)>>,
+    ) -> Result<ReclockIter<'a, M>, (PartitionId, MzOffset)> {
+        let mut batch_upper = HashMap::with_capacity(batch.len());
+        for (pid, messages) in batch.iter_mut() {
+            messages.sort_unstable_by(|a, b| a.1.cmp(&b.1));
+            if let Some((_msg, offset)) = messages.first() {
+                let part_since = self.partition_since(pid);
+                if !(part_since <= *offset) {
+                    return Err((pid.clone(), *offset));
```
[the remark this answers is not in the document]
Comment (aljoscha) [reply src/storage/src/source/reclock.rs:139]: :+1: I like the notion that you need to compact in `SourceTime` to use compacted bindings

**Review by cjubb39 — COMMENTED** [review]
couple (quick!) inline qs.  After that, I think should be ready to go!

File: src/storage/src/source/reclock.rs (hunk 16 above)
[the remark this answers is not in the document]
Comment (cjubb39) [reply src/storage/src/source/reclock.rs:148]: > this extra bit of code wouldn't add much to ensuring global progress

Right -- to be clear, my argument doesn't concern global progress.  My argument is that the more sure we are that an individual replica will always make progress, the easier it is to reason about the correctness of the system.   It's not an argument about correctness (I think the code is correct!) but one about code readability.

"we have N replicas that, no matter what, will make progress so we know we must in fact make global progress" is more straightforward than "individual replicas may or may not make progress and could in fact get stuck in (albeit rare) loops but we have convinced ourselves that those situations won't turn into an issue of not making global progress".  You're losing an amount of modularity -- which I think is the biggest factor in people whether people can understand and reason about the system enough to confidently make changes

File: src/storage/src/source/reclock.rs (hunk 17 above)
[the remark this answers is not in the document]
Comment (petrosagg) [reply src/storage/src/source/reclock.rs:388]: Having messages reclocked *at* the `as_of` is not possible. In order to allow it I would have to remove the safety feature if refusing to reclock messages that are in the past (and might be reclocked incorrectly). The compaction/as_of test that is in place already tests what happens at the boundary

File: src/storage/src/source/reclock.rs (hunk 16 above)
[the remark this answers is not in the document]
Comment (petrosagg) [reply src/storage/src/source/reclock.rs]: I still disagree that this would lead to easier to understand logic, for me if statements are always better if they don't exist. 

In order to resolve this conversation I created new `advance` function that loops forever until the upper is advanced as dictated by `next_mint_timestamp` and made `reclock` only add to the remap collection when bindings need to be minted. These are now called at the appropriate times by the source ingestion code. I put the diff in a separate commit https://github.com/MaterializeInc/materialize/pull/13123/commits/1688ba78f8b0d662fb64c1b7e1ce11860e47a0cb

File: src/storage/src/source/reclock.rs [hunk 24]
```diff
@@ -8,413 +8,795 @@
...
+        let reclocked_msgs = operator
+            .reclock(&mut batch)
+            .await
+            .unwrap()
+            .flat_map(|(_, msgs)| msgs)
+            .collect_vec();
+        assert_eq!(reclocked_msgs, &[(1, 0), (2, 0), (3, 1000), (4, 1000)]);
+        assert!(batch[&PART_ID].is_empty());
+
+        // Reclock a batch that contains multiple messages having the same offset
```
[the remark this answers is not in the document]
Comment (petrosagg) [reply src/storage/src/source/reclock.rs:735]: sure, changed

**Review by cjubb39 — APPROVED** [review]
looks great! thanks for all the back and forth!

**Comment by bosconi** [comment]
bugbot run
